internal/identity/providers.go

```diff
@@ -188,6 +191,47 @@ func (p *Provider) IdentityFromToken(ctx context.Context, t *oauth2.Token) (*oid
 
 // Revoke enables a user to revoke her token. If the identity provider supports revocation
 // the endpoint is available, otherwise an error is thrown.
+//
+// We will attempt to get the identity provider's possible revocation endpoint by making a
+// request to their /.well-known/openid-configuration, and based on the common endpoint keys
+// we'll be using the revocation_endpoint and end_session_endpoint keys  to get the
+// endpoint and the implement it in the most generic way.
+//
+// Google : https://accounts.google.com/.well-known/openid-configuration
+// Azure: https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration
 func (p *Provider) Revoke(ctx context.Context, token *oauth2.Token) error {
-	return fmt.Errorf("internal/identity: revoke not implemented by %s", p.ProviderName)
+	var response struct {
+		RevocationURL string `json:"revocation_endpoint"`
+		EndSessionURL string `json:"end_session_endpoint"`
+	}
+
+	// try to see if the provider has a way to revoke token
+	wellKnownURL := p.ProviderURL + "/.well-known/openid-configuration"
+	err := httputil.Client(ctx, http.MethodGet, wellKnownURL, version.UserAgent(), nil, nil, &response)
```

Comment:
Instead of doing an additional call to the well known endpoint, we should already have this information when we do 
```
	// build api client to make group membership api calls
	if err := p.provider.Claims(&gp); err != nil {
		return nil, err
	}
``` 
```go
// Claims unmarshals raw fields returned by the server during discovery.

//    var claims struct {
//        ScopesSupported []string `json:"scopes_supported"`
//        ClaimsSupported []string `json:"claims_supported"`
//    }

//    if err := provider.Claims(&claims); err != nil {
//        // handle unmarshaling error
//    }

// For a list of fields defined by the OpenID Connect spec see:
// https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
```
on provider instantiation. Though the claim name (as you have here) will vary by provider, I don't think it makes sense to re-query the configuration endpoint for every revoke request.

We should be able to pass your response struct to the oidc `Provider.Claims` method to return the desired endpoint.

Reply:
That's right, thanks for the feedback, I will work on it.